src/pymor/vectorarrays/block.py

```diff
@@ -21,11 +21,23 @@ class BlockVectorArray(VectorArrayInterface):
     :class:`~pymor.operators.block.BlockOperator`.
     """
 
+    _dtype_invalid = True # force setting dtype on first access
+
     def __init__(self, blocks, space):
         self._blocks = tuple(blocks)
         self.space = space
         assert self._blocks_are_valid()
 
+    @property
+    def dtype(self):
+        if self._dtype_invalid:
+            if len(self._blocks): # else defaults to base class
+                self._cached_dtype = reduce(np.promote_types, (v.dtype for v in self._blocks))
```

Comment:
How expensive is this? Does it need to be cached?

Reply:
My rationale was that `dtype` is checked very often, but mutated rarely.

Reply:
Ok.